### PR TITLE
[flink] add flink compaction job of unaware-bucket compaction

### PR DESCRIPTION
**Pull request by leaves12138**
Standalone compaction job for unaware bucket table (unaware bucket append-only mode table)

add test for Batch and Streaming mode compaction job

### CODE REVIEW

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkConnectorOptions.java [hunk 1]
```diff
@@ -87,6 +87,15 @@ public class FlinkConnectorOptions {
                                     + "for each statement individually by also considering the global configuration. "
                                     + "If user enable the scan.infer-parallelism, the planner will derive the parallelism by inferred parallelism.");
 
+    public static final ConfigOption<Integer> COMPACTION_PARALLELISM =
+            ConfigOptions.key("compaction.parallelism")
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkConnectorOptions.java:91]: Do you want to use it in many places? Or is it just for the unaware table? If it's just a standalone use, I suggest using the key starting with unaware

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java [hunk 2]
```diff
@@ -82,6 +83,10 @@ public List<AppendOnlyCompactionTask> run() {
         return Collections.emptyList();
     }
 
+    public void withFilter(Predicate filter) {
+        scan.withFilter(filter);
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java:87]: the filter should be passed in the constructor.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionTopologyBuilder.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+/**
+ * Build for unaware-bucket table flink compaction job.
+ *
+ * <p>Note: This compaction job class is only used for unaware-bucket compaction, at start-up, it
+ * scans all the files from the latest snapshot, filter large file, and add small files into memory,
+ * generates compaction task for them. At continuous, it scans the delta files from the follow-up
+ * snapshot. We need to enable checkpoint for this compaction job, checkpoint will trigger committer
+ * stage to commit all the compacted files.
+ */
+public class UnawareBucketCompactionTopologyBuilder implements Serializable {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionTopologyBuilder.java:63]: `UnawareBucketCompactTopoBuilder`, make this shorter...

Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionTopologyBuilder.java:63]: Why need to be a Serializable

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionTopologyBuilder.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        // from source, construct the full flink job
+        sinkFromSource(source);
+    }
+
+    private DataStreamSource<AppendOnlyCompactionTask> buildSource() {
+        UnawareBucketSourceFunction source = new UnawareBucketSourceFunction(table);
+        long scanInterval = table.coreOptions().continuousDiscoveryInterval().toMillis();
+        source.withStreaming(isContinuous);
+        source.withScanInterval(scanInterval);
+        source.withFilter(getPartitionFilter());
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionTopologyBuilder.java:106]: UnawareBucketSourceFunction don't need to a builder pattern, parameters can be passed by constructor.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionTopologyBuilder.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+                        COMPACTION_WORKER_NAME + " -> " + table.name(),
+                        committableTypeInfo,
+                        new AppendOnlyTableCompactionWorkerOperator(table, commitUser));
+
+        if (compactionWorkerParallelism != null) {
+            compacted.setParallelism(compactionWorkerParallelism);
+        }
+
+        // commit stage, to commit compaction messages
+        SingleOutputStreamOperator<?> committed =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionTopologyBuilder.java:147]: Can you create a subclass of `FlinkSink`?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionTopologyBuilder.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        boolean streamingCheckpointEnabled =
+                isContinuous && checkpointConfig.isCheckpointingEnabled();
+        if (streamingCheckpointEnabled) {
+            FlinkSink.assertCheckpointConfiguration(env);
+        }
+
+        // compaction worker stage, to execute compaction tasks
+        CommittableTypeInfo committableTypeInfo = new CommittableTypeInfo();
+        SingleOutputStreamOperator<Committable> compacted =
+                input.transform(
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionTopologyBuilder.java:137]: No `PartitionTransformation`? Default is forward?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/SimpleVersionedCompactionTaskSerializer.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import org.apache.paimon.append.AppendOnlyCompactionTask;
+import org.apache.paimon.table.sink.CompactionTaskSerializer;
+
+import org.apache.flink.core.io.SimpleVersionedSerializer;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+/** {@link SimpleVersionedSerializer} for {@link AppendOnlyCompactionTask}. */
+public class SimpleVersionedCompactionTaskSerializer
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/SimpleVersionedCompactionTaskSerializer.java:30]: Just `CompactionTaskSimpleSerializer`

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/UnawareBucketSourceFunction.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        compactionCoordinator = new AppendOnlyTableCompactionCoordinator(table);
+        if (filter != null) {
+            compactionCoordinator.withFilter(filter);
+        }
+    }
+
+    @Override
+    public void run(SourceContext<AppendOnlyCompactionTask> sourceContext) throws Exception {
+        // scan loop
+        do {
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/UnawareBucketSourceFunction.java:54]: Take a look to `MonitorFunction`.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionTopologyBuilder.java (hunk 6 above)
Comment (leaves12138) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionTopologyBuilder.java]: Seems we don't need specified partitioner. Any AppendOnlyCompactionTask could be invoke by any Compaction Worker

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/UnawareBucketSourceFunction.java [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        compactionCoordinator = new AppendOnlyTableCompactionCoordinator(table, streaming, filter);
+    }
+
+    @Override
+    public void run(SourceContext<AppendOnlyCompactionTask> sourceContext) throws Exception {
+        while (isRunning) {
+            boolean isEmpty;
+            try {
+                // do scan and plan action, emit append-only compaction tasks.
+                List<AppendOnlyCompactionTask> tasks = compactionCoordinator.run();
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/source/UnawareBucketSourceFunction.java:76]: need a lock here.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/UnawareBucketCompactionSink.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            AppendOnlyFileStoreTable table,
+            Integer compactionWorkerParallelism,
+            DataStream<AppendOnlyCompactionTask> input) {
+        return new UnawareBucketCompactionSink(table, compactionWorkerParallelism).sinkFrom(input);
+    }
+
+    @Override
+    protected OneInputStreamOperator<AppendOnlyCompactionTask, Committable> createWriteOperator(
+            StoreSinkWrite.Provider writeProvider, boolean isStreaming, String commitUser) {
+        return null;
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/UnawareBucketCompactionSink.java:73]: Can we also implements this? Just ignore `writeProvider`

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionTopologyBuilder.java (hunk 6 above)
Comment (JingsongLi) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionTopologyBuilder.java]: We should set rebalance by hard code.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionJob.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+                        table, isContinuous, scanInterval, getPartitionFilter());
+
+        return UnawareBucketSourceFunction.buildSource(env, source, isContinuous, tableIdentifier);
+    }
+
+    private void sinkFromSource(DataStreamSource<AppendOnlyCompactionTask> input) {
+        Options conf = Options.fromMap(table.options());
+        Integer compactionWorkerParallelism =
+                conf.get(FlinkConnectorOptions.UNAWARE_BUCKET_COMPACTION_PARALLELISM);
+        PartitionTransformation<AppendOnlyCompactionTask> transformation =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/compact/UnawareBucketCompactionJob.java:97]: input.rebalance is OK

**Review by JingsongLi — APPROVED** [review]

**Comment by JingsongLi** [comment]
+1

**Pull request merged**